File: Sources/Wispry/TextPipeline.swift
import Foundation

struct ProcessedText {
    var text: String
    var shouldPressEnter: Bool
    var cancelled: Bool
}

enum TextPipeline {
    static func process(_ rawText: String, style: TransformStyle, snippets: [VoiceSnippet]) -> ProcessedText {
        var text = normalizeWhitespace(rawText)
        guard !text.isEmpty else {
            return ProcessedText(text: "", shouldPressEnter: false, cancelled: false)
        }

        if isCancelCommand(text) {
            return ProcessedText(text: "", shouldPressEnter: false, cancelled: true)
        }

        var shouldPressEnter = false
        let enterCommands = ["press enter", "hit enter", "send it"]
        for command in enterCommands where endsWithCommand(text, command) {
            text = removeTrailingCommand(text, command)
            shouldPressEnter = true
            break
        }

        text = expandSnippets(in: text, snippets: snippets)
        let commandStyle = styleFromSpokenCommand(&text)
        text = applySelfCorrections(text)
        text = normalizeSpeechArtifacts(text)
        text = replaceDictationPhrases(text)
        text = normalizePunctuationSpacing(text)
        text = removeConversationalLeadIns(text)
        text = removeFillers(text)
        text = collapseRepeatedClauseRevisions(text)
        text = removeRepeatedWords(text)

        let finalStyle = commandStyle ?? style
        text = apply(style: finalStyle, to: text)
        text = normalizePunctuationSpacing(text)
        text = normalizeWhitespace(text)

        return ProcessedText(text: text, shouldPressEnter: shouldPressEnter, cancelled: false)
    }

    private static func normalizeWhitespace(_ text: String) -> String {
        text
            .replacingOccurrences(of: "\u{00a0}", with: " ")
            .replacingOccurrences(of: #"[ \t]+"#, with: " ", options: .regularExpression)
            .replacingOccurrences(of: #"\n{3,}"#, with: "\n\n", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func isCancelCommand(_ text: String) -> Bool {
        let lowered = text.lowercased().trimmingCharacters(in: .punctuationCharacters.union(.whitespacesAndNewlines))
        return ["cancel", "cancel that", "discard that", "stop listening"].contains(lowered)
    }

    private static func endsWithCommand(_ text: String, _ command: String) -> Bool {
        text.lowercased()
            .trimmingCharacters(in: .punctuationCharacters.union(.whitespacesAndNewlines))
            .hasSuffix(command)
    }

    private static func removeTrailingCommand(_ text: String, _ command: String) -> String {
        guard let range = text.range(of: command, options: [.caseInsensitive, .backwards]) else { return text }
        return String(text[..<range.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines.union(.punctuationCharacters))
    }

    private static func expandSnippets(in input: String, snippets: [VoiceSnippet]) -> String {
        var text = input
        for snippet in snippets where !snippet.phrase.isEmpty {
            if text.compare(snippet.phrase, options: [.caseInsensitive, .diacriticInsensitive]) == .orderedSame {
                return snippet.expansion
            }
            text = text.replacingOccurrences(
                of: snippet.phrase,
                with: snippet.expansion,
                options: [.caseInsensitive, .diacriticInsensitive]
            )
        }
        return text
    }

    private static func styleFromSpokenCommand(_ text: inout String) -> TransformStyle? {
        let commands: [(String, TransformStyle)] = [
            ("make this more professional", .professional),
            ("make this professional", .professional),
            ("rewrite this professionally", .professional),
            ("make this casual", .casual),
            ("turn this into a list", .list),
            ("turn to list", .list),
            ("format as a list", .list),
            ("leave this verbatim", .verbatim)
        ]

        for (command, style) in commands {
            guard text.lowercased().hasPrefix(command) else { continue }
            text = String(text.dropFirst(command.count))
                .trimmingCharacters(in: .whitespacesAndNewlines.union(.punctuationCharacters))
            return style
        }
        return nil
    }

    private static func replaceDictationPhrases(_ input: String) -> String {
        var text = input
        let replacements: [(phrase: String, replacement: String, allowedAtStart: Bool, blockedNextWords: Set<String>)] = [
            ("new paragraph", "\n\n", true, []),
            ("new line", "\n", true, []),
            ("open quote", "\"", true, []),
            ("close quote", "\"", true, []),
            ("quote", "\"", false, ["unquote", "mark", "marks"]),
            ("single quote", "'", true, []),
            ("open parenthesis", "(", true, []),
            ("close parenthesis", ")", true, []),
            ("open bracket", "[", true, []),
            ("close bracket", "]", true, []),
            ("question mark", "?", false, ["icon", "symbol"]),
            ("exclamation mark", "!", false, ["icon", "symbol"]),
            ("exclamation point", "!", false, ["icon", "symbol"]),
            ("comma", ",", false, ["separated", "delimited", "splice", "operator"]),
            ("period", ".", false, ["of", "piece", "pain", "drama", "costume", "tracking"]),
            ("full stop", ".", false, []),
            ("colon", ":", false, ["cancer", "health", "screening"]),
            ("semicolon", ";", false, ["insertion"]),
            ("dash", "-", false, ["board", "cam", "camera", "lane"]),
            ("slash", "/", false, ["command", "commands", "fiction", "mark"]),
            ("at sign", "@", true, [])
        ]

        for replacement in replacements {
            text = replaceCommandPhrase(
                in: text,
                phrase: replacement.phrase,
                with: replacement.replacement,
                allowedAtStart: replacement.allowedAtStart,
                blockedNextWords: replacement.blockedNextWords
            )
        }
        return text
    }

    private static func replaceCommandPhrase(
        in input: String,
        phrase: String,
        with replacement: String,
        allowedAtStart: Bool,
        blockedNextWords: Set<String>
    ) -> String {
        let escaped = NSRegularExpression.escapedPattern(for: phrase)
        let pattern = #"(?i)(^|\s)"# + escaped + #"(?=\s|$|[.,!?;:])"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return input }

        var output = input
        let nsInput = input as NSString
        let matches = regex.matches(in: input, range: NSRange(location: 0, length: nsInput.length))
        for match in matches.reversed() {
            let leadingRange = match.range(at: 1)
            let phraseStart = match.range.location + leadingRange.length
            let phraseLength = match.range.length - leadingRange.length
            let previousText = nsInput.substring(to: phraseStart).trimmingCharacters(in: .whitespacesAndNewlines)
            if !allowedAtStart && previousText.isEmpty {
                continue
            }
            if let next = nextWord(after: phraseStart + phraseLength, in: input),
               blockedNextWords.contains(next.lowercased()) {
                continue
            }

            output = (output as NSString).replacingCharacters(
                in: NSRange(location: phraseStart, length: phraseLength),
                with: replacement
            )
        }
        return output
    }

    private static func nextWord(after location: Int, in input: String) -> String? {
        let nsInput = input as NSString
        guard location < nsInput.length else { return nil }
        let rest = nsInput.substring(from: location)
        guard let regex = try? NSRegularExpression(pattern: #"^\s*([A-Za-z]+)"#),
              let match = regex.firstMatch(in: rest, range: NSRange(location: 0, length: (rest as NSString).length)),
              match.range(at: 1).location != NSNotFound else {
            return nil
        }
        return (rest as NSString).substring(with: match.range(at: 1))
    }

    private static func normalizeSpeechArtifacts(_ input: String) -> String {
        input
            .replacingOccurrences(of: #"(?i)(?:\s|^)(?:\[\s*blank[_ ]audio\s*\]|\(\s*blank[_ ]audio\s*\)|<\|nospeech\|>)(?=\s|$)"#, with: " ", options: .regularExpression)
            .replacingOccurrences(of: #"(?i)\bet cetera\b"#, with: "etc.", options: .regularExpression)
            .replacingOccurrences(of: #"(?i)\betcetera\b"#, with: "etc.", options: .regularExpression)
    }

    private static func normalizePunctuationSpacing(_ input: String) -> String {
        input
            .replacingOccurrences(of: #"\s+([,.;:!?])"#, with: "$1", options: .regularExpression)
            .replacingOccurrences(of: #"([,;:!?])([^\s\d,.;:!?])"#, with: "$1 $2", options: .regularExpression)
            .replacingOccurrences(of: #"(?<!\d)\.([^\s\d,.;:!?])"#, with: ". $1", options: .regularExpression)
            .replacingOccurrences(of: #"\betc\s*\."#, with: "etc.", options: [.regularExpression, .caseInsensitive])
            .replacingOccurrences(of: #"\s{2,}"#, with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func applySelfCorrections(_ input: String) -> String {
        let strongPatterns = [
            #"(?i)\b(?:take that back|i take that back|scratch that|ignore that|discard that|forget that|remove that)\b[,:;\-\s]*"#,
            #"(?i)\b(?:actually i meant|i didn't mean|i did not mean|what i meant was|no[, ]+i mean|sorry[, ]+i mean|no[, ]+actually)\b[,:;\-\s]*"#
        ]

        for pattern in strongPatterns {
            if let replacement = textAfterLastMatch(pattern: pattern, in: input, requirePriorText: false) {
                return replacement
            }
        }

        if let replacement = textAfterLastMatch(
            pattern: #"(?i)\bi mean\b[,:;\-\s]*"#,
            in: input,
            requirePriorText: true
        ) {
            return replacement
        }

        return input
    }

    private static func textAfterLastMatch(pattern: String, in input: String, requirePriorText: Bool) -> String? {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
        let nsInput = input as NSString
        let matches = regex.matches(in: input, range: NSRange(location: 0, length: nsInput.length))
        guard let match = matches.last else { return nil }

        let prefix = nsInput.substring(to: match.range.location).trimmingCharacters(in: .whitespacesAndNewlines.union(.punctuationCharacters))
        if requirePriorText && prefix.isEmpty {
            return nil
        }

        let start = match.range.location + match.range.length
        guard start < nsInput.length else { return nil }
        let replacement = nsInput.substring(from: start).trimmingCharacters(in: .whitespacesAndNewlines.union(.punctuationCharacters))
        return replacement.count > 1 ? replacement : nil
    }

    private static func removeFillers(_ input: String) -> String {
        input
            .replacingOccurrences(of: #"\b(um|uh|erm|ah|hmm)\b,?\s*"#, with: "", options: .regularExpression)
            .replacingOccurrences(of: #"\b(kind of|sort of|you know|you know what i mean)\b,?\s*"#, with: "", options: [.regularExpression, .caseInsensitive])
            .replacingOccurrences(of: #"\s{2,}"#, with: " ", options: .regularExpression)
    }

    private static func removeConversationalLeadIns(_ input: String) -> String {
        input
            .replacingOccurrences(
                of: #"(?i)^\s*((also|and)\s+)?(yeah|yep|okay|ok|right|so|well)\b[,\s]*"#,
                with: "",
                options: .regularExpression
            )
            .replacingOccurrences(
                of: #"(?i)^\s*(also|and)\b[,\s]+"#,
                with: "",
                options: .regularExpression
            )
            .replacingOccurrences(of: #"\s{2,}"#, with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func removeRepeatedWords(_ input: String) -> String {
        var previous = ""
        var output: [String] = []

        for rawWord in input.split(separator: " ") {
            let word = String(rawWord)
            let normalized = word.lowercased().trimmingCharacters(in: .punctuationCharacters)
            if normalized != previous || normalized.isEmpty {
                output.append(word)
            }
            previous = normalized
        }

        return output.joined(separator: " ")
    }

    private static func collapseRepeatedClauseRevisions(_ input: String) -> String {
        let words = input.split(separator: " ").map(String.init)
        guard words.count >= 8 else { return input }

        for split in 2..<(words.count - 2) {
            guard split <= 24 else { break }
            let left = Array(words[..<split])
            let right = Array(words[split...])
            guard right.count >= 4, left.count <= 24 else { continue }
            guard tokenKey(left[0]) == tokenKey(right[0]),
                  tokenKey(left[1]) == tokenKey(right[1]) else {
                continue
            }

            let comparedRight = Array(right.prefix(left.count))
            guard revisionSimilarity(left: left, right: comparedRight) >= 0.72 else {
                continue
            }

            return right.joined(separator: " ")
        }

        return input
    }

    private static func revisionSimilarity(left: [String], right: [String]) -> Double {
        guard !left.isEmpty, !right.isEmpty else { return 0 }
        var rightCounts: [String: Int] = [:]
        for word in right {
            let key = tokenKey(word)
            guard !key.isEmpty else { continue }
            rightCounts[key, default: 0] += 1
        }

        var shared = 0
        for word in left {
            let key = tokenKey(word)
            guard let count = rightCounts[key], count > 0 else { continue }
            shared += 1
            rightCounts[key] = count - 1
        }

        return Double(shared) / Double(max(left.count, right.count))
    }

    private static func tokenKey(_ input: String) -> String {
        input
            .lowercased()
            .trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
    }

    private static func apply(style: TransformStyle, to input: String) -> String {
        switch style {
        case .verbatim:
            return input.trimmingCharacters(in: .whitespacesAndNewlines)
        case .clean:
            return finishSentence(capitalizeSentences(input))
        case .professional:
            return makeProfessional(input)
        case .casual:
            return makeCasual(input)
        case .list:
            return makeList(input)
        }
    }

    private static func capitalizeFirst(_ input: String) -> String {
        guard let first = input.first else { return input }
        return first.uppercased() + input.dropFirst()
    }

    private static func capitalizeSentences(_ input: String) -> String {
        var output = ""
        var shouldCapitalize = true

        for character in input {
            if shouldCapitalize, character.isLetter {
                output.append(contentsOf: character.uppercased())
                shouldCapitalize = false
                continue
            }

            output.append(character)
            if ".!?\n".contains(character) {
                shouldCapitalize = true
            } else if !character.isWhitespace {
                shouldCapitalize = false
            }
        }

        return output
    }

    private static func finishSentence(_ input: String) -> String {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let last = trimmed.last else { return trimmed }
        if ".!?:;)]}".contains(last) || trimmed.contains("\n") {
            return trimmed
        }
        return trimmed + (looksLikeQuestion(trimmed) ? "?" : ".")
    }

    private static func looksLikeQuestion(_ input: String) -> Bool {
        let firstWord = input
            .lowercased()
            .split(whereSeparator: { !$0.isLetter })
            .first
            .map(String.init)

        guard let firstWord else { return false }
        let questionWords: Set<String> = [
            "who", "what", "when", "where", "why", "how",
            "is", "are", "am", "was", "were",
            "do", "does", "did",
            "can", "could", "will", "would", "should",
            "has", "have", "had"
        ]
        return questionWords.contains(firstWord)
    }

    private static func makeProfessional(_ input: String) -> String {
        var text = capitalizeSentences(input)
        text = text.replacingOccurrences(of: #"(?i)\bhey\b"#, with: "Hello", options: .regularExpression)
        text = text.replacingOccurrences(of: #"(?i)\bthanks\b"#, with: "Thank you", options: .regularExpression)
        text = text.replacingOccurrences(of: #"(?i)\bi think\b"#, with: "I think", options: .regularExpression)
        return finishSentence(text)
    }

    private static func makeCasual(_ input: String) -> String {
        var text = capitalizeSentences(input)
        text = text.replacingOccurrences(of: #"(?i)\bhello\b"#, with: "Hey", options: .regularExpression)
        return finishSentence(text)
    }

    private static func makeList(_ input: String) -> String {
        let pieces = input
            .replacingOccurrences(of: " and then ", with: ". ", options: .caseInsensitive)
            .replacingOccurrences(of: " then ", with: ". ", options: .caseInsensitive)
            .components(separatedBy: CharacterSet(charactersIn: ".;\n"))
            .map { normalizeWhitespace($0) }
            .filter { !$0.isEmpty }

        guard !pieces.isEmpty else { return input }
        return pieces.map { "- " + capitalizeFirst($0) }.joined(separator: "\n")
    }
}

File: windows/src/text-pipeline.js
const TransformStyle = Object.freeze({
  verbatim: "Verbatim",
  clean: "Clean",
  professional: "Professional",
  casual: "Casual",
  list: "List"
});

function processText(rawText, style = TransformStyle.clean, snippets = []) {
  let text = normalizeWhitespace(rawText);
  if (!text) {
    return { text: "", shouldPressEnter: false, cancelled: false };
  }

  if (isCancelCommand(text)) {
    return { text: "", shouldPressEnter: false, cancelled: true };
  }

  let shouldPressEnter = false;
  for (const command of ["press enter", "hit enter", "send it"]) {
    if (endsWithCommand(text, command)) {
      text = removeTrailingCommand(text, command);
      shouldPressEnter = true;
      break;
    }
  }

  text = expandSnippets(text, snippets);
  const commandStyle = styleFromSpokenCommand(text);
  text = commandStyle.text;
  const preservationSource = text;
  text = applySelfCorrections(text);
  text = normalizeSpeechArtifacts(text);
  text = replaceDictationPhrases(text);
  text = normalizePunctuationSpacing(text);
  text = removeConversationalLeadIns(text);
  text = removeFillers(text);
  text = collapseRepeatedClauseRevisions(text);
  text = removeRepeatedWords(text);

  const finalStyle = commandStyle.style || style;
  text = applyStyle(finalStyle, text);
  text = normalizePunctuationSpacing(text);
  text = normalizeWhitespace(text);
  if (!preservesEnoughContent(text, preservationSource)) {
    text = conservativeCleanup(preservationSource, finalStyle);
  }

  return { text, shouldPressEnter, cancelled: false };
}

function normalizeWhitespace(input) {
  return String(input || "")
    .replace(/\u00a0/g, " ")
    .replace(/[ \t]+/g, " ")
    .replace(/\n{3,}/g, "\n\n")
    .trim();
}

function conservativeCleanup(input, style) {
  let text = normalizeSpeechArtifacts(input);
  text = replaceDictationPhrases(text);
  text = normalizePunctuationSpacing(text);
  text = removeConversationalLeadIns(text);
  text = removeFillers(text);
  text = removeRepeatedWords(text);
  text = applyStyle(style, text);
  text = normalizePunctuationSpacing(text);
  return normalizeWhitespace(text);
}

function preservesEnoughContent(processed, original) {
  const originalCount = wordCount(original);
  if (originalCount < 35) {
    return true;
  }
  const processedCount = wordCount(processed);
  return processedCount >= Math.max(12, Math.floor(originalCount * 0.55));
}

function wordCount(input) {
  return String(input || "").split(/[^A-Za-z0-9]+/).filter(Boolean).length;
}

function stripCommandText(input) {
  return String(input || "")
    .toLowerCase()
    .replace(/^[\s.,!?;:]+|[\s.,!?;:]+$/g, "");
}

function isCancelCommand(text) {
  return ["cancel", "cancel that", "discard that", "stop listening"].includes(stripCommandText(text));
}

function endsWithCommand(text, command) {
  return stripCommandText(text).endsWith(command);
}

function removeTrailingCommand(text, command) {
  const index = text.toLowerCase().lastIndexOf(command.toLowerCase());
  if (index < 0) {
    return text;
  }
  return text.slice(0, index).replace(/[\s.,!?;:]+$/g, "");
}

function expandSnippets(input, snippets) {
  let text = input;
  for (const snippet of snippets || []) {
    const phrase = String(snippet.phrase || "").trim();
    const expansion = String(snippet.expansion || "");
    if (!phrase) {
      continue;
    }
    if (text.localeCompare(phrase, undefined, { sensitivity: "accent" }) === 0) {
      return expansion;
    }
    text = text.replace(new RegExp(escapeRegExp(phrase), "gi"), expansion);
  }
  return text;
}

function styleFromSpokenCommand(input) {
  const commands = [
    ["make this more professional", TransformStyle.professional],
    ["make this professional", TransformStyle.professional],
    ["rewrite this professionally", TransformStyle.professional],
    ["make this casual", TransformStyle.casual],
    ["turn this into a list", TransformStyle.list],
    ["turn to list", TransformStyle.list],
    ["format as a list", TransformStyle.list],
    ["leave this verbatim", TransformStyle.verbatim]
  ];

  const lowered = input.toLowerCase();
  for (const [command, style] of commands) {
    if (lowered.startsWith(command)) {
      return {
        style,
        text: input.slice(command.length).replace(/^[\s.,!?;:]+/g, "")
      };
    }
  }
  return { style: null, text: input };
}

function replaceDictationPhrases(input) {
  const replacements = [
    ["new paragraph", "\n\n", true, []],
    ["new line", "\n", true, []],
    ["open quote", "\"", true, []],
    ["close quote", "\"", true, []],
    ["quote", "\"", false, ["unquote", "mark", "marks"]],
    ["single quote", "'", true, []],
    ["open parenthesis", "(", true, []],
    ["close parenthesis", ")", true, []],
    ["open bracket", "[", true, []],
    ["close bracket", "]", true, []],
    ["question mark", "?", false, ["icon", "symbol"]],
    ["exclamation mark", "!", false, ["icon", "symbol"]],
    ["exclamation point", "!", false, ["icon", "symbol"]],
    ["comma", ",", false, ["separated", "delimited", "splice", "operator"]],
    ["period", ".", false, ["of", "piece", "pain", "drama", "costume", "tracking"]],
    ["full stop", ".", false, []],
    ["colon", ":", false, ["cancer", "health", "screening"]],
    ["semicolon", ";", false, ["insertion"]],
    ["dash", "-", false, ["board", "cam", "camera", "lane"]],
    ["slash", "/", false, ["command", "commands", "fiction", "mark"]],
    ["at sign", "@", true, []]
  ];

  let text = input;
  for (const [phrase, replacement, allowedAtStart, blockedNextWords] of replacements) {
    text = replaceCommandPhrase(text, phrase, replacement, allowedAtStart, new Set(blockedNextWords));
  }
  return text;
}

function replaceCommandPhrase(input, phrase, replacement, allowedAtStart, blockedNextWords) {
  const regex = new RegExp(`(^|\\s)${escapeRegExp(phrase)}(?=\\s|$|[.,!?;:])`, "gi");
  const matches = Array.from(input.matchAll(regex));
  let output = input;

  for (const match of matches.reverse()) {
    const leadingLength = match[1].length;
    const start = match.index + leadingLength;
    const end = start + phrase.length;
    const previousText = input.slice(0, start).trim();
    if (!allowedAtStart && !previousText) {
      continue;
    }
    const next = nextWord(input, end);
    if (next && blockedNextWords.has(next.toLowerCase())) {
      continue;
    }
    output = output.slice(0, start) + replacement + output.slice(end);
  }

  return output;
}

function nextWord(input, location) {
  const match = input.slice(location).match(/^\s*([A-Za-z]+)/);
  return match ? match[1] : null;
}

function normalizeSpeechArtifacts(input) {
  return input
    .replace(/(?:\s|^)(?:\[\s*blank[_ ]audio\s*\]|\(\s*blank[_ ]audio\s*\)|<\|nospeech\|>)(?=\s|$)/gi, " ")
    .replace(/\bet cetera\b/gi, "etc.")
    .replace(/\betcetera\b/gi, "etc.");
}

function normalizePunctuationSpacing(input) {
  return input
    .replace(/\s+([,.;:!?])/g, "$1")
    .replace(/([,;:!?])([^\s\d,.;:!?])/g, "$1 $2")
    .replace(/(^|[^\d])\.([^\s\d,.;:!?])/g, "$1. $2")
    .replace(/\betc\s*\./gi, "etc.")
    .replace(/\s{2,}/g, " ")
    .trim();
}

function applySelfCorrections(input) {
  const strongPatterns = [
    /\b(?:take that back|i take that back|scratch that|ignore that|discard that|forget that|remove that)\b[,:;\-\s]*/gi,
    /\b(?:actually i meant|i didn't mean|i did not mean|what i meant was|no[, ]+i mean|sorry[, ]+i mean|no[, ]+actually)\b[,:;\-\s]*/gi
  ];

  for (const pattern of strongPatterns) {
    const replacement = textAfterLastMatch(pattern, input, false);
    if (replacement) {
      return replacement;
    }
  }

  return textAfterLastMatch(/\bi mean\b[,:;\-\s]*/gi, input, true) || input;
}

function textAfterLastMatch(pattern, input, requirePriorText) {
  const matches = Array.from(input.matchAll(pattern));
  const match = matches.at(-1);
  if (!match) {
    return null;
  }
  const prefix = input.slice(0, match.index).replace(/^[\s.,!?;:]+|[\s.,!?;:]+$/g, "");
  if (requirePriorText && !prefix) {
    return null;
  }
  const start = match.index + match[0].length;
  const replacement = input.slice(start).replace(/^[\s.,!?;:]+|[\s.,!?;:]+$/g, "");
  return replacement.length > 1 ? replacement : null;
}

function removeFillers(input) {
  return input
    .replace(/\b(um|uh|erm|ah|hmm)\b,?\s*/gi, "")
    .replace(/\b(kind of|sort of|you know|you know what i mean)\b,?\s*/gi, "")
    .replace(/\s{2,}/g, " ");
}

function removeConversationalLeadIns(input) {
  return input
    .replace(/^\s*((also|and)\s+)?(yeah|yep|okay|ok|right|so|well)\b[,\s]*/gi, "")
    .replace(/^\s*(also|and)\b[,\s]+/gi, "")
    .replace(/\s{2,}/g, " ")
    .trim();
}

function removeRepeatedWords(input) {
  let previous = "";
  const output = [];
  for (const word of input.split(" ")) {
    const normalized = tokenKey(word);
    if (normalized !== previous || !normalized) {
      output.push(word);
    }
    previous = normalized;
  }
  return output.join(" ");
}

function collapseRepeatedClauseRevisions(input) {
  const words = input.split(" ");
  if (words.length < 8) {
    return input;
  }

  for (let split = 2; split < words.length - 2 && split <= 24; split += 1) {
    const left = words.slice(0, split);
    const right = words.slice(split);
    if (right.length < 4 || left.length > 24) {
      continue;
    }
    if (tokenKey(left[0]) !== tokenKey(right[0]) || tokenKey(left[1]) !== tokenKey(right[1])) {
      continue;
    }
    if (revisionSimilarity(left, right.slice(0, left.length)) >= 0.72) {
      return right.join(" ");
    }
  }

  return input;
}

function revisionSimilarity(left, right) {
  if (!left.length || !right.length) {
    return 0;
  }

  const counts = new Map();
  for (const word of right) {
    const key = tokenKey(word);
    if (key) {
      counts.set(key, (counts.get(key) || 0) + 1);
    }
  }

  let shared = 0;
  for (const word of left) {
    const key = tokenKey(word);
    const count = counts.get(key) || 0;
    if (count > 0) {
      shared += 1;
      counts.set(key, count - 1);
    }
  }
  return shared / Math.max(left.length, right.length);
}

function tokenKey(input) {
  return String(input || "").toLowerCase().replace(/^[^a-z0-9]+|[^a-z0-9]+$/gi, "");
}

function applyStyle(style, input) {
  switch (style) {
    case TransformStyle.verbatim:
      return input.trim();
    case TransformStyle.professional:
      return makeProfessional(input);
    case TransformStyle.casual:
      return makeCasual(input);
    case TransformStyle.list:
      return makeList(input);
    case TransformStyle.clean:
    default:
      return finishSentence(capitalizeSentences(input));
  }
}

function capitalizeSentences(input) {
  let output = "";
  let shouldCapitalize = true;
  for (const character of input) {
    if (shouldCapitalize && /\p{L}/u.test(character)) {
      output += character.toUpperCase();
      shouldCapitalize = false;
      continue;
    }
    output += character;
    if (".!?\n".includes(character)) {
      shouldCapitalize = true;
    } else if (!/\s/.test(character)) {
      shouldCapitalize = false;
    }
  }
  return output;
}

function finishSentence(input) {
  const trimmed = input.trim();
  if (!trimmed) {
    return trimmed;
  }
  const last = trimmed.at(-1);
  if (".!?:;)]}".includes(last) || trimmed.includes("\n")) {
    return trimmed;
  }
  return trimmed + (looksLikeQuestion(trimmed) ? "?" : ".");
}

function looksLikeQuestion(input) {
  const firstWord = input.toLowerCase().split(/[^a-z]+/).find(Boolean);
  return new Set([
    "who", "what", "when", "where", "why", "how",
    "is", "are", "am", "was", "were",
    "do", "does", "did",
    "can", "could", "will", "would", "should",
    "has", "have", "had"
  ]).has(firstWord);
}

function makeProfessional(input) {
  return finishSentence(
    capitalizeSentences(input)
      .replace(/\bhey\b/gi, "Hello")
      .replace(/\bthanks\b/gi, "Thank you")
      .replace(/\bi think\b/gi, "I think")
  );
}

function makeCasual(input) {
  return finishSentence(capitalizeSentences(input).replace(/\bhello\b/gi, "Hey"));
}

function makeList(input) {
  const pieces = input
    .replace(/ and then /gi, ". ")
    .replace(/ then /gi, ". ")
    .split(/[.;\n]/)
    .map(normalizeWhitespace)
    .filter(Boolean);
  if (!pieces.length) {
    return input;
  }
  return pieces.map((piece) => `- ${capitalizeFirst(piece)}`).join("\n");
}

function capitalizeFirst(input) {
  return input ? input[0].toUpperCase() + input.slice(1) : input;
}

function escapeRegExp(input) {
  return String(input).replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

module.exports = {
  TransformStyle,
  processText
};
